parcels/kernel.py

```diff
@@ -405,28 +409,35 @@ def remove_deleted(pset):
         remove_deleted(pset)
 
         # Identify particles that threw errors
-        # NOTE: this relies on the SOA implementation, it does not
-        # generalize. This needs to be reworked.
-        error_particles = np.isin(pset.collection._data['state'], [StateCode.Success, StateCode.Evaluate], invert=True)
-        while np.any(error_particles):
+        while True:
+            n_error = 0  # Number of particles with a different code.
             # Apply recovery kernel
-            for p in np.where(error_particles)[0]:
-                particles = pset.collection.get_single_by_index(p)
-                if particles.state == OperationCode.StopExecution:
+            # Effectively loop over all particles that don't have sucess of
+            # evaluate codes.
+            for particle in pset:
+                if particle.state in [StateCode.Evaluate, StateCode.Success]:
+                    continue
+
+                if particle.state == OperationCode.StopExecution:
                     return
-                if particles.state == OperationCode.Repeat:
-                    particles.set_state(StateCode.Evaluate)
-                elif particles.state in recovery_map:
-                    recovery_kernel = recovery_map[particles.state]
-                    particles.set_state(StateCode.Success)
-                    recovery_kernel(particles, self.fieldset, particles.time)
-                    if particles.state == StateCode.Success:
-                        particles.set_state(StateCode.Evaluate)
+                if particle.state == OperationCode.Repeat:
+                    particle.set_state(StateCode.Evaluate)
+                elif particle.state in recovery_map:
+                    recovery_kernel = recovery_map[particle.state]
+                    particle.set_state(StateCode.Success)
+                    recovery_kernel(particle, self.fieldset, particle.time)
+                    if particle.state == StateCode.Success:
+                        particle.set_state(StateCode.Evaluate)
                 else:
-                    logger.warning_once('Deleting particle because of bug in #749 and #737')
-                    particles.delete()
+                    logger.warning_once(
+                        'Deleting particle because of bug in #749 and #737')
+                    particle.delete()
```

Comment:
I actually recently fixed that one in my benchmarking branch. It's a simple fix, so I may just add that right here.

The issue is: if the kernel sets the particle state already to 'Delete', this if-branch is executed and hence emits the error - although everything behaves just-as-planned.